### ISSUE TITLE
sshd: service not restarted after config change

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#43341]
## Issue description

config was not active until i run `systemctl restart sshd`.

maybe this is intentional, so ssh connections are not disconncted? in such a case there should be a hint for the user to restart manually!

### Steps to reproduce

1. edit config `nano /etc/nixos/configuration.nix`
2. `nixos-rebuild switch`
```
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/h9d1wkgww4794nxxbpsp81m5yicvq308-etc-sshd_config.drv
  /nix/store/0pf7g2fdal4m760gfmazr8h9ma3ylrpd-etc.drv
  /nix/store/k8wl8gfxfkgbnvk8m9sy4ggkcdsz1n78-nixos-system-nixos-18.09pre145524.2a8a5533d18.drv
building '/nix/store/h9d1wkgww4794nxxbpsp81m5yicvq308-etc-sshd_config.drv'...
building '/nix/store/0pf7g2fdal4m760gfmazr8h9ma3ylrpd-etc.drv'...
building '/nix/store/k8wl8gfxfkgbnvk8m9sy4ggkcdsz1n78-nixos-system-nixos-18.09pre145524.2a8a5533d18.drv'...
updating GRUB 2 menu...
activating the configuration...
setting up /etc...
setting up tmpfiles
```

3. `systemctl status sshd`
```
● sshd.service - SSH Daemon
   Loaded: loaded (/nix/store/z6i9a6k47bl1dh371vasj1ri6yyq4qha-unit-sshd.service/sshd.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2018-07-11 10:21:25 UTC; 51min ago
 Main PID: 661 (sshd)
    Tasks: 1 (limit: 2385)
   Memory: 6.9M
   CGroup: /system.slice/sshd.service
           └─661 /nix/store/amyv4cr6kmbwpdj9f1v2gyyv68yj546q-openssh-7.7p1/bin/sshd -f /etc/ssh/sshd_config
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.53, NixOS, 18.09pre145524.2a8a5533d18 (Jellyfish)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.4`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by davidak** [comment]
This is not a problem on 18.03!

```
[root@nixos:~]# nixos-rebuild switch --upgrade
unpacking channels...
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/svq6xgma07iyh4a81pd1dbp9kfxr157j-sshd_config.drv
  /nix/store/v81gss9fw0csy13s9nkgg5j6d0x2r6v9-unit-sshd.service.drv
  /nix/store/2kvv91jj9l3nbkaikr7kcl7wwi3f58wk-system-units.drv
  /nix/store/wiyks7563c1wdnzk9qf35afrm0ifa5wl-etc.drv
  /nix/store/3lbf44kpyh7b3iwd467c91sff1k9wikj-nixos-system-ethmoid-18.03.132847.aec217852f2.drv
building '/nix/store/svq6xgma07iyh4a81pd1dbp9kfxr157j-sshd_config.drv'...
building '/nix/store/v81gss9fw0csy13s9nkgg5j6d0x2r6v9-unit-sshd.service.drv'...
building '/nix/store/2kvv91jj9l3nbkaikr7kcl7wwi3f58wk-system-units.drv'...
building '/nix/store/wiyks7563c1wdnzk9qf35afrm0ifa5wl-etc.drv'...
building '/nix/store/3lbf44kpyh7b3iwd467c91sff1k9wikj-nixos-system-ethmoid-18.03.132847.aec217852f2.drv'...
activating the configuration...
setting up /etc...
setting up tmpfiles
restarting the following units: sshd.service
the following new units were started: plymouth-quit.service
```
Maybe because here also the service unit changes.

```
[root@nixos:~]# systemctl status sshd.service
● sshd.service - SSH Daemon
   Loaded: loaded (/nix/store/rvd9b30rp4vw1n497ccz52y17xxz7sk9-unit-sshd.service/sshd.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2018-07-11 14:36:50 CEST; 2min 22s ago
  Process: 30237 ExecStartPre=/nix/store/cjifjq1j5zcfc4xq4mfm1cn682hhznfc-unit-script/bin/sshd-pre-start (code=exited, status=0/SUCCESS)
 Main PID: 30239 (sshd)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/sshd.service
           └─30239 /nix/store/dk3hbw2lvhq4y3fgli1vjzk77fklj7n8-openssh-7.6p1/bin/sshd -f /nix/store/j4fp1vv9s0szhh0mxsmjlxd4zsn68qa8-sshd_config
```
The `sshd_config` used here has a path directly to nix-store. In master it is `/etc/ssh/sshd_config` (which is a symlink into nix store).

**Comment by nixy** [comment]
Related #39118

IMO the best behavior would be to not restart the service and issue a warning indicating that it needs to be restarted manually. But that's also a big pain and may not work well on systems doing unattended upgrades. Maybe a config option to specify this behavior?

**Comment by peterhoeg** [comment]
It's probably due to ```services.openssh.startWhenNeeded```. When ssh is running in socket activated mode, the individually spawned instances are not restarted (and shouldn't be).

**Comment by davidak** [comment]
@peterhoeg so new connections would use the new config? that would be great.
I don't use that mode currently. Should i?

@nixy maybe it should restart when all connections are closed? then it would cause no problems.

I would not use unattended upgrades, but when i think about this situation, what about an option to reboot the system when necessary (after kernel or glibc update) in that mode? Maybe i create an issue to suggest that.

**Comment by peterhoeg** [comment]
> so new connections would use the new config?

Correct

> I don't use that mode currently. Should i?

In my opinion everything should be at least socket activated as it immediately solves all ordering issues for us and then the activated service should either run permanently (for high volume services) and spawned once per connection (for low volume).

**Comment by volth** [comment]
Still reproducible.

Adding a new port to ```services.openssh.ports``` following ```nixos-rebuild switch``` does not restart/reload ```sshd```, so it does not start listening on a new port (combining with firewall, which by default makes a hole for ```services.openssh.ports``` and which **does** restart, it could result in a unrechable server - firewall makes a hole for a new port while ```sshd``` is still listening on the old one)

**Comment by volth** [comment]
Also, ```systemctl restart sshd``` does not abort active connections

**Comment by fpletz** [comment]
Fixed by #53050.
